src/webServiceClient.ts

```diff
@@ -0,0 +1,153 @@
+import https = require('https');
+import mmdb = require('maxmind');
+import * as models from './models';
+
+type servicePaths = 'city' | 'country' | 'insights';
+
+/** Class representing the WebServiceClient **/
+export default class WebServiceClient {
+  private accountID: string;
+  private licenseKey: string;
+  private timeout: number;
+  private host = 'geoip.maxmind.com';
+
+  /**
+   * Instanstiates a WebServiceClient
+   *
+   * @param accountID The account ID
+   * @param licenseKey The license key
+   */
+  public constructor(accountID: string, licenseKey: string, timeout?: number) {
+    this.accountID = accountID;
+    this.licenseKey = licenseKey;
+    this.timeout = timeout || 3000;
+  }
+
+  /**
+   * Returns a Promise with the City Precision data for an IP address.
+   *
+   * @param ipAddress The IP Address you want to query the City webservice with
+   */
+  public city(ipAddress: string): Promise<models.City> {
+    return this.responseFor<models.City>('city', ipAddress, models.City);
+  }
+
+  /**
+   * Returns a Promise with the Country Precision data for an IP address.
+   *
+   * @param ipAddress The IP Address you want to query the Country webservice with
+   */
+  public country(ipAddress: string): Promise<models.Country> {
+    return this.responseFor<models.Country>(
+      'country',
+      ipAddress,
+      models.Country
+    );
+  }
+
+  /**
+   * Returns a Promise with the Insights Precision data for an IP address.
+   *
+   * @param ipAddress The IP Address you want to query the Insights webservice with
+   */
+  public insights(ipAddress: string): Promise<models.Insights> {
+    return this.responseFor<models.Insights>(
+      'insights',
+      ipAddress,
+      models.Insights
+    );
+  }
+
+  private responseFor<T>(
+    path: servicePaths,
+    ipAddress: string,
+    modelClass: any
+  ): Promise<T> {
+    if (!mmdb.validate(ipAddress)) {
+      return Promise.reject({
+        code: 'IP_ADDRESS_INVALID',
+        error: 'The IP address provided is invalid',
+      });
+    }
+
+    const parsedPath = `/geoip/v2.1/${path}/${ipAddress}`;
```

Comment:
We should probably have a way to specify which version of the API to use, and default to the latest. This can allow users to lock in on a version for whatever reason they may have.

Reply:
I disagree with this suggestion for a couple of reasons:

* Users should be concerned with what version of the client API they're using, not which version of the API endpoint the client API is using.
* If a breaking endpoint API version is released, a corresponding breaking client API version would be released as well.  This relieves us of needing to provide backward compatibility for endpoint versions

Reply:
Good points. I now concur.